### PR TITLE
Feature rabbit configurator

### PR DESCRIPTION
**Pull request by IdoRozval**

### CODE REVIEW

File: Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+using EasyNetQ;
+using EasyNetQ.Topology;
+
+namespace Blizzard.IO.RabbitMQ
+{
+    public class NetqRabbitConfigurator : IRabbitConfigurator
+    {
+        private readonly IBus _netqBus;
+        private readonly IDictionary<string, IExchange> _nameToNetqExchange;
+        private readonly IDictionary<string, IQueue> _nameToNetqQueue;
```
Comment (alloncm) [inline Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs]: Why those are at the class level and not the function level?
you have no use for them after the function has done and if youll call the function again youll clear them anyway.

File: Blizzard.IO.RabbitMQ/Helpers.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿using Blizzard.IO.RabbitMQ.Entities;
+using System.Collections.Generic;
+
+namespace Blizzard.IO.RabbitMQ
+{
+    public static class Helpers
```
Comment (alloncm) [inline Blizzard.IO.RabbitMQ/Helpers.cs]: Myabe make this class internal.
I dont think youll even call this in the tests
```suggestion
    internal static class Helpers
```

Comment (alloncm) [reply Blizzard.IO.RabbitMQ/Helpers.cs]: Maybe call this class in a better name I think even Utilities is better.
```suggestion
    public static class Utilities
```
if you accept it dont forget to change the file name

File: Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        {
+            _nameToNetqExchange.Clear();
+            _nameToNetqQueue.Clear(); 
+            DeclareExchanges(configuration.Exchanges);
+            DeclareQueues(configuration.Queues);
+            DeclareExchangeToQueueBindings(configuration.ExchangeToQueueBindings);
+            DeclareExchangeToExchangeBindings(configuration.ExchangeToExchangeBindings);
+        }
+
+        private void DeclareExchangeToExchangeBindings(List<RabbitBinding> exchangeToExchangeBindings)
```
Comment (alloncm) [inline Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs]: Also you can change the list you are passing in all the functions to IReadOnlyCollection
```suggestion
        private void DeclareExchangeToExchangeBindings(IReadOnlyCollection<RabbitBinding> exchangeToExchangeBindings)
```

**Review by alloncm — COMMENTED** [review]
Hi I did a little CR I got tired of my tests
@ybs007  if you dont like my comments please comment and tell me.

File: Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs (hunk 1 above)
Comment (IdoRozval) [reply Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs]: I Agree

File: Blizzard.IO.RabbitMQ/Helpers.cs (hunk 2 above)
Comment (IdoRozval) [reply Blizzard.IO.RabbitMQ/Helpers.cs]: I need those in the tests too

**Review by ybs007 — COMMENTED** [review]
Add tests

File: Blizzard.IO.RabbitMQ/Helpers.cs (hunk 2 above)
Comment (IdoRozval) [reply Blizzard.IO.RabbitMQ/Helpers.cs]: I didn't change the access modifier to `internal` but I renamed the class to `Utilities`

File: Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs (hunk 3 above)
Comment (IdoRozval) [reply Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs]: I refactored the class and it turns out that using all of these private methods made the class very ugly, so i put everything in the same method, it look better than before

Comment (alloncm) [reply Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs]: I think this function is much harder to read now.
Can you explain why did you removed those private function?

File: Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+using Microsoft.Extensions.Logging;
+
+namespace Blizzard.IO.RabbitMQ
+{
+    public class NetqRabbitConfigurator : IRabbitConfigurator
+    {
+        private readonly IBus _netqBus;
+        private readonly ILogger<NetqRabbitConfigurator> _logger;
+
+        public NetqRabbitConfigurator(IBus netqBus, ILogger<NetqRabbitConfigurator> logger)
```
Comment (alloncm) [inline Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs]: Pass a logger factory like you did in the consumer

File: Blizzard.IO.RabbitMQ.Tests/NetqRabbitConfiguratorTests.cs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    public class NetqRabbitConfiguratorTests
+    {
+        private RabbitConfiguration _configuration;
+        private RabbitExchange _exchange;
+        private RabbitQueue _queue;
+
+        private Mock<IBus> _netqBusMock;
+        private Mock<IAdvancedBus> _netqAdvancedBusMock;
+        private Mock<ILogger<NetqRabbitConfigurator>> _loggerMock;
+        private NetqRabbitConfigurator _netqRabbitConfigurator;
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ.Tests/NetqRabbitConfiguratorTests.cs]: Add space between the Mocks and tested class

File: Blizzard.IO.RabbitMQ.Tests/NetqRabbitConfiguratorTests.cs [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+                Queues = new List<RabbitQueue>(),
+                ExchangeToExchangeBindings = new List<RabbitBinding>(),
+                ExchangeToQueueBindings = new List<RabbitBinding>()
+            };
+
+            _loggerMock = new Mock<ILogger<NetqRabbitConfigurator>>();
+            _netqAdvancedBusMock = new Mock<IAdvancedBus>();
+            _netqBusMock = new Mock<IBus>();
+            _netqBusMock.SetupGet(bus => bus.Advanced).Returns(_netqAdvancedBusMock.Object);
+            _netqRabbitConfigurator = new NetqRabbitConfigurator(_netqBusMock.Object, _loggerMock.Object);
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ.Tests/NetqRabbitConfiguratorTests.cs]: Add space between mock initialization and setups and between setup to tested class initialization

File: Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs (hunk 3 above)
Comment (IdoRozval) [reply Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs]: כ

File: Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs (hunk 4 above)
Comment (IdoRozval) [reply Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs]: f

File: Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+{
+    public class NetqRabbitConfigurator : IRabbitConfigurator
+    {
+        private readonly IBus _netqBus;
+        private readonly ILogger _logger;
+
+        public NetqRabbitConfigurator(IBus netqBus, ILoggerFactory loggerFactory)
+        {
+            _netqBus = netqBus;
+            _logger = loggerFactory.CreateLogger(typeof(NetqRabbitConfigurator));
```
Comment (alloncm) [inline Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs]: Use the non static override of this funciton (the one that receives the string)

Comment (IdoRozval) [reply Blizzard.IO.RabbitMQ/NetqRabbitConfigurator.cs]: f

**Pull request merged**